### PR TITLE
Improve declarative API annotations

### PR DESCRIPTION
**Pull request by layday**
### Description
Improves the declarative API's annotations.

Some legacy constructs are not typed.

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [x] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

### CODE REVIEW

File: sqlalchemy-stubs/orm/decl_api.pyi [hunk 1]
```diff
@@ -25,13 +34,10 @@ class DeclarativeMeta(type):
     def __setattr__(cls, key: Any, value: Any) -> None: ...
     def __delattr__(cls, key: Any) -> None: ...
     metadata: MetaData
-    registry: "registry"
+    registry: _registry  # Avoid circural reference
```
Comment (zzzeek) [inline sqlalchemy-stubs/orm/decl_api.pyi:37]: how does this part work?  how is "_registry" different from "registry" ?

File: sqlalchemy-stubs/orm/decl_api.pyi [hunk 2]
```diff
@@ -46,42 +52,51 @@ class _stateful_declared_attr(declared_attr):
...
+    def as_declarative_base(
+        self, **kw: Any
+    ) -> Callable[[Type[_T]], Type[_T]]: ...
     def map_declaratively(self, cls: type) -> Mapper: ...
     def map_imperatively(
-        self, class_: Any, local_table: Optional[Any] = ..., **kw: Any
-    ): ...
+        self,
+        class_: type,
+        local_table: Optional[Union[FromClause, Table]] = ...,
```
Comment (zzzeek) [inline sqlalchemy-stubs/orm/decl_api.pyi:96]: a Table is a FromClause

File: sqlalchemy-stubs/orm/decl_api.pyi [hunk 3]
```diff
@@ -46,42 +52,51 @@ class _stateful_declared_attr(declared_attr):
...
     @property
-    def mappers(self): ...
+    def mappers(self) -> FrozenSet[Mapper]: ...
     def configure(self, cascade: bool = ...) -> None: ...
     def dispose(self, cascade: bool = ...) -> None: ...
     def generate_base(
         self,
-        mapper: Optional[Any] = ...,
-        cls: Any = ...,
+        mapper: Optional[Mapper] = ...,
```
Comment (zzzeek) [inline sqlalchemy-stubs/orm/decl_api.pyi:83]: this is a Callable that would return Mapper objects.   i dont even know why we have this but it's still there.

File: sqlalchemy-stubs/orm/decl_api.pyi [hunk 4]
```diff
@@ -46,42 +52,51 @@ class _stateful_declared_attr(declared_attr):
...
 def declarative_base(
-    bind: Optional[Any] = ...,
-    metadata: Optional[Any] = ...,
-    mapper: Optional[Any] = ...,
-    cls: Any = ...,
+    bind: Optional[
+        Any
+    ] = ...,  # NOTE: Deprecated in 1.4, to be removed in 2.0.
+    metadata: Optional[MetaData] = ...,
+    mapper: Optional[Mapper] = ...,
```
Comment (zzzeek) [inline sqlalchemy-stubs/orm/decl_api.pyi:59]: callable that returns Mapper objects

**Review by zzzeek — CHANGES_REQUESTED** [review]
couple of adjustments

File: sqlalchemy-stubs/orm/decl_api.pyi (hunk 1 above)
Comment (layday) [reply sqlalchemy-stubs/orm/decl_api.pyi:37]: "registry" refers to `DeclarativeMeta.registry` which is a circural reference.

**Comment by zzzeek** [comment]
graet!  thanks

**Pull request merged**